### PR TITLE
remove Future from StreamRefs mat val, #24372

### PR DESCRIPTION
**Pull request by patriknw**
## Purpose

Remove the Future wrapping of SourceRef and SinkRef.

## References

References #24372

## Changes

* use the new `eagerMaterializer` parameter to be able to access the stage ActorRef immediately when the stage logic is created (same approach as in `ActorRefSource`)
* a few lazy could also be eliminated thanks to this

## Background Context

StreamRefs was marked with ApiMayChange for the reason that the Future wasn't great to have in the API. With the new `getEagerStageActor` it was possible to get rid of the Future.

### CODE REVIEW

**Review by johanandren — APPROVED** [review]
Nice
LGTM!

**Review by chbatey — APPROVED** [review]
LGTM, great. Don't know why Mima didn't pick this up.

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala [hunk 1]
```diff
@@ -43,24 +50,30 @@ private[stream] final class SinkRefStageImpl[In] private[akka] (val initialPartn
       case OptionVal.None      => "<no-initial-ref>"
     }
 
-  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
-    val promise = Promise[SourceRefImpl[In]]
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, SourceRef[In]) =
+    throw new IllegalStateException("Not supported")
```
Comment (raboof) [inline akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: Can a user ever trigger this? If so, can we tell them what to do when they encounter this problem?

**Review by raboof — APPROVED** [review]
Looks a lot nicer!

**Comment by helena** [comment]
Waiting for this to merge until completing https://github.com/akka/akka/pull/26838

File: akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala (hunk 1 above)
Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/streamref/SinkRefImpl.scala]: Not in any reasonable scenario AFAICS

File: akka-stream/src/main/mima-filters/2.5.x.backwards.excludes [hunk 2]
```diff
@@ -60,4 +60,7 @@ ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowO
 ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.zipLatestGraph")
 ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.zipLatest")
 ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.zipLatestWithGraph")
-ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.zipLatestWith")
\ No newline at end of file
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.zipLatestWith")
+
+# #24372 No Future/CompletionStage in StreamRefs
+# FIXME why was change not detected?
```
Comment (patriknw) [inline akka-stream/src/main/mima-filters/2.5.x.backwards.excludes]: Added separate ticket for this. Pretty serious so we should look into it soonish. https://github.com/akka/akka/issues/26852

**Pull request merged**